### ISSUE TITLE
Use in SSR requires extra check to ensure `media` isn't null

### ISSUE THREAD

**Issue by diverent2** [pearofducks/svelte-match-media#5]
Hello there :)

I've run into a problem with Sapper and the module.
Since Sapper doesn't have a "main.js" and I need the breakpoints globally, I've figured I should init `setup()` inside the `_layout.svelte` component.
The good news: This works fine in all the components that follow inside `<slot />` 👍  (maybe this should get added to the docs)

However I have a globally used `Navigation` component and it doesn't want to accept the `setup()` from `_layout.svelte`

The problem seems to be that it can't subscribe to the `window` as window is undefined when the component first gets initialized.

`window is not defined`.

Basically this is what this looks like

> _layout.svelte
```
<script>
  import Navigation from "./../components/Navigation/Navigation.svelte";

    import { setup } from 'svelte-match-media';

  setup({
  medium_down: 'screen and (max-width: 600px)',
  medium_up: 'screen and (min-width: 601px)'
})

  export let segment;

</script>

<main>
  <Navigation {segment} />      👎 this context doesn't work
  <slot />                     👍 this context works fine
</main>

```

Any idea what to do about this? 🤔 
I really want to use the plugin inside the navigation, it would optimize things by a lot.

**Comment by pearofducks** [comment]
In your `_layout.svelte` you can ensure the code runs only on the client in a few ways - [this issue](https://github.com/sveltejs/sapper/issues/222) is old, but looks like you can use the `oncreate` hook and set up things there?

---

You might also be able to conditionally add `setup` in `client.js`.

If it seems like it'd be helpful for `media` to be able to take a default in SSR-land instead of `undefined` let me know, that seems like it might be useful.

---

Sorry I don't have any direct answers, I haven't had the chance to use Sapper yet!

**Comment by pearofducks** [comment]
@diverent2 - I was able to test the above ideas in Sapper.

If you run `setup` in `client.js`, you can freely use `media`, as long as you ensure media isn't null first - since on the server it won't be initialized yet.

e.g. 

```
{#if $media && $media.desktop}
  <h1>I am on desktop</h1>
{/if}
```

I think this extra check is sub-optimal, so I'm going to keep this issue and look into an improvement - but might change the title to be more descriptive of the problem that needs solving.

**Comment by diverent2** [comment]
Hi there,
I've did some testing but sadly my two main problems remain:

First things first:

Running `setup` in `client.js` before `sapper.start()` works for me too and is better then doing this inside `_layout.svelte` imo.

However, this actually doesn't really solve my two issues:

## Using media queries in sapper throws `error 500` for a moment.

See #7

## The original issue seems to continue as I still can't use `media` in a component inside `_layout.svelte`

As the title states, the original issue remains.

`client.js`
```js
import * as sapper from '@sapper/app';
import { setup } from 'svelte-match-media';

setup();

sapper.start({
  target: document.querySelector('#sapper'),
});
```

`_layout.svelte`
```html
<script>
  import Navigation from "Navigation.svelte";
  import { media } from "svelte-match-media";
</script>

<main>
  <Navigation />
  <slot />
</main>
```

`Navigation.svelte`
```html
<script>
  import { media } from "svelte-match-media";
</script>
  {#if $media && $media.desktop}
    <h1>I am on desktop</h1>
  {/if}
```

Still throws 
> Cannot read property 'subscribe' of undefined


Any ideas? 🤷

**Comment by pearofducks** [comment]
Closing this after deeper investigation.

Svelte seems to get in its own way quite a bit here with regards to how it handles stores and the lifecycle.

Stores seem to be attached before the `onMount` hook runs - meaning we'd need to provide _some_ initial store-ish value to make the client happy. However, when this value is then updated after mounting, Svelte refuses to re-read the new variable, and maintains it's subscription to the original pre-mount store.

This makes it so this problem is only really addressed through a component, which needlessly sets up multiple matchMedia listeners per-component.

If there are alternatives in the future, please raise them.
